syne_tune/optimizer/schedulers/searchers/searcher.py

```diff
@@ -470,3 +472,113 @@ def clone_from_state(self, state: dict):
+        else:
+            assert isinstance(debug_log, DebugLogPrinter)
+            self._debug_log = debug_log
+
+    def get_config(self, **kwargs):
+        """Sample a new configuration at the grid of config_space
+        This is done without replacement, so previously returned configs are
+        not suggested again.
+
+        Returns
+        -------
+        A new configuration that is valid, or None if no new config can be
+        suggested.
+        """
+        trial_id = kwargs.get("trial_id")
+        if self._debug_log is not None:
+            self._debug_log.start_get_config("grid", trial_id=trial_id)
+        new_config = self._next_initial_config()
+        if new_config is None:
+            while True:
+                _config = self._next_config_candidate()
+                if _config is None:
+                    # If searcher returns no config at all, the
+                    # search space is exhausted
+                    break
+                if not self._excl_list.contains(_config):
+                    new_config = _config
+                    break
+
+        if new_config is not None:
+            # Add config to exclusion list so that it won't be suggested again
+            self._excl_list.add(new_config)
+            if self._debug_log is not None:
+                self._debug_log.set_final_config(new_config)
+                # All get_config debug log info is only written here
+                self._debug_log.write_block()
+        else:
+            # Fail to get
+            msg = (
+                "All the configuration has already been chosen. "
+                + f"Exclusion list has size {len(self._excl_list)}."
+            )
+            cs_size = self._excl_list.configspace_size
+            if cs_size is not None:
+                msg += f" Configuration space has size {cs_size}."
+            logger.warning(msg)
+
+        return new_config
+
+    def _generate_config_candidates(self, config_space: Dict) -> List[Dict]:
+        # Validates and get all hp values
+        hp_values = []
+        for hp_range in config_space.values():
+            assert isinstance(hp_range, Categorical)
+            hp_values.append(hp_range.categories)
+
+        # Generate all combinations of hp values
+        hp_values_combinations = list(product(*hp_values))
+
+        # Map each hp value with its name and append to grids_to_evaluate
+        keys = config_space.keys()
+        grids_to_evaluate = []
+        for values in hp_values_combinations:
+            grids_to_evaluate.append(dict(zip(keys, values)))
+        return grids_to_evaluate
+
+    def _next_config_candidate(self) -> Optional[Dict]:
+        if self._config_candidates:
+            return self._config_candidates.pop(0)
+        else:
+            return None  # No more grids
+
+    def get_state(self) -> dict:
+        state = dict(
+            super().get_state(),
+            config_candidates=self._config_candidates,
```

Comment:
Would be nice not to have to store this in the state. Can you not recreate it, say by just storing its size?